### PR TITLE
EOS:17889 Implement cluster start for 32 nodes 

### PR DESCRIPTION
**Pull request by VijayThakkar132**
Signed-off-by: Vijaykumar Thakkar <vijaykumar.thakkar@seagate.com>

## Problem Statement
<pre>
  <code>
  Story Ref (if any):
    EOS:17889 Provide cluster start for 32 nodes
  </code>
</pre>
## Unit testing on RPM done
<pre>
  <code>
  Yes
  </code>
</pre>
## Problem Description
<pre>
  <code>
    Implement cluster start operation to support till 32 nodes
  </code>
</pre>
## Solution
<pre>
  <code>
    Implemented cluster start operation
  </code>
</pre>
## Unit Test Cases
<pre>
  <code>
    Verified cluster start on the 5 nodes VM and it went well.
  </code>
</pre>

### CODE REVIEW

File: ha/core/controllers/cluster_controller.py [hunk 1]
```diff
@@ -28,6 +28,14 @@ def __init__(self):
         """
         super(ClusterController, self).__init__()
 
+    @controller_error_handler
```
Comment (ajay-paratmandali) [inline ha/core/controllers/cluster_controller.py:31]: For all class no need to add 
`@controller_error_handler`
It is only used if function is exposed

File: ha/core/controllers/pcs/cluster_controller.py [hunk 2]
```diff
@@ -38,7 +48,37 @@ def start(self) -> dict:
...
+                        _res = self._controllers[const.NODE_CONTROLLER].start(_node_id)
+                        _res = json.loads(_res)
+                        if _res.get("status") == const.STATUSES.FAILED.value:
+                            msg = _res.get("msg")
+                            Log.error(f"Node {_node_id} : {msg}")
+                    time.sleep(30)
+
+                return {"status": const.STATUSES.IN_PROGRESS.value, "msg": "Cluster start operation performed"}
+            else:
+                return {"status": const.STATUSES.FAILED.value, "msg": "Node list is empty"}
```
Comment (ajay-paratmandali) [inline ha/core/controllers/pcs/cluster_controller.py:78]: for node list failed msg
msg: 
Cluster start failed. Not able to verify node list.

File: ha/core/controllers/cluster_controller.py (hunk 1 above)
Comment (VijayThakkar132) [reply ha/core/controllers/cluster_controller.py:31]: Yes, Not needed. I will remove it.

File: ha/core/controllers/pcs/cluster_controller.py [hunk 3]
```diff
@@ -38,7 +48,37 @@ def start(self) -> dict:
...
+        _, _, _rc = self._execute.run_cmd(const.PCS_CLUSTER_STATUS, check_error=False)
+        if _rc != 0:
+            self._execute.run_cmd(const.PCS_CLUSTER_START, check_error=False)
+            time.sleep(60)
+            _, _, _rc = self._execute.run_cmd(const.PCS_CLUSTER_STATUS, check_error=False)
+            if _rc != 0:
+                return {"status": const.STATUSES.FAILED.value, "msg": "Cluster start operation failed"}
+
+        _res = self.node_list()
+        _res = json.loads(_res)
```
Comment (ajay-paratmandali) [inline ha/core/controllers/pcs/cluster_controller.py:60]: `_res`

add nodelist output in log info

File: ha/core/controllers/pcs/cluster_controller.py (hunk 2 above)
Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py:79]: Done

File: ha/core/controllers/pcs/cluster_controller.py (hunk 3 above)
Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py:60]: Added log to print node list when _res returns success

File: ha/core/cluster/cluster_manager.py [hunk 4]
```diff
@@ -283,6 +283,9 @@ def __init__(self):
             # Example: cm.cluster_controller.start()
             # Find more example in test case.
             self.__dict__[controller] = self._controllers[controller]
+        for controller in self._controllers.keys():
```
Comment (ajay-paratmandali) [inline ha/core/cluster/cluster_manager.py:286]: Can we move this in ElementControllerFactory
as it is part of class build

File: ha/core/controllers/pcs/cluster_controller.py (hunk 2 above)
Comment (ajay-paratmandali) [reply ha/core/controllers/pcs/cluster_controller.py:79]: Missed one change

`{"status": const.STATUSES.FAILED.value, "msg": "Node list is empty"}`

Comment (ajay-paratmandali) [reply ha/core/controllers/pcs/cluster_controller.py:79]: (optional)
Also for any error you can raise
ClusterManagerError

@controller_error_handler will covert error to json

File: ha/core/cluster/cluster_manager.py (hunk 4 above)
Comment (VijayThakkar132) [reply ha/core/cluster/cluster_manager.py]: Moved

File: ha/core/controllers/pcs/cluster_controller.py (hunk 2 above)
Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py]: Done

**Review by ajay-paratmandali — APPROVED** [review]

File: ha/core/controllers/pcs/cluster_controller.py [hunk 5]
```diff
@@ -38,7 +48,38 @@ def start(self) -> dict:
...
+                _node_group = [ _node_list[i:i + const.PCS_NODE_START_GROUP_SIZE] for i in range(0, len(_node_list), const.PCS_NODE_START_GROUP_SIZE)]
+
+                for _node_subgroup in _node_group:
+                    for _node_id in _node_subgroup:
+                        _res = self._controllers[const.NODE_CONTROLLER].start(_node_id)
+                        _res = json.loads(_res)
+                        if _res.get("status") == const.STATUSES.FAILED.value:
+                            msg = _res.get("msg")
+                            Log.error(f"Node {_node_id} : {msg}")
+                    time.sleep(30)
```
Comment (indrajitzagade) [inline ha/core/controllers/pcs/cluster_controller.py:75]: How did we decide on 30 seconds here?

File: ha/core/controllers/pcs/cluster_controller.py [hunk 6]
```diff
@@ -38,7 +48,38 @@ def start(self) -> dict:
...
+
+                return {"status": const.STATUSES.IN_PROGRESS.value, "msg": "Cluster start operation performed"}
+            else:
+                return {"status": const.STATUSES.FAILED.value, "msg": "Cluster start failed. Not able to verify node list."}
+        else:
+            return {"status": _res.get("status"), "msg": _res.get("msg")}
```
Comment (ajaysrivas) [inline ha/core/controllers/pcs/cluster_controller.py:81]: The start operation should have it's own message. The list_node message can be appended to that.

File: ha/core/controllers/pcs/cluster_controller.py [hunk 7]
```diff
@@ -38,7 +48,38 @@ def start(self) -> dict:
             ([dict]): Return dictionary. {"status": "", "msg":""}
                 status: Succeeded, Failed, InProgress
         """
-        raise HAUnimplemented("Cluster start operation is not implemented.")
+        _, _, _rc = self._execute.run_cmd(const.PCS_CLUSTER_STATUS, check_error=False)
+        if _rc != 0:
```
Comment (ajaysrivas) [inline ha/core/controllers/pcs/cluster_controller.py:52]: where is the code for starting cluster on this node?

File: ha/core/controllers/pcs/cluster_controller.py (hunk 5 above)
Comment (ajaysrivas) [reply ha/core/controllers/pcs/cluster_controller.py:75]: Put this timeout in conf file. Make it less, may be 5 seconds and retry for 30 seconds if that is required. This base timeout should be used for all pcs operations. Retries in conf will be specific to operations.

Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py:75]: Actually, the Main intention of this delay is to wait till all the resources get started for the sub-group. So, I felt 30 secs in enough. Later, We can remove the delay and wait till all the resources become online.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: ha/core/controllers/pcs/cluster_controller.py (hunk 5 above)
Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py:75]: Sure. I will modify it as per Ajay's suggestion

Comment (indrajitzagade) [reply ha/core/controllers/pcs/cluster_controller.py:75]: Sure

Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py]: As discussed with Ajay S, modified the implementation.

File: ha/core/cluster/cluster_manager.py [hunk 8]
```diff
@@ -209,7 +209,7 @@ def start(self):
                 raise Exception("Cluster failed to start; pcs not installed")
             # if cluster is not running; start cluster
             elif(_err.find("cluster is not currently running on this node") != -1):
-                self._execute.run_cmd(const.PCS_CLUSTER_START, check_error=False)
+                self._execute.run_cmd(const.PCS_CLUSTER_START_ALL, check_error=False)
```
[the remark this answers is not in the document]
Comment (VijayThakkar132) [reply ha/core/cluster/cluster_manager.py]: Done

File: ha/core/controllers/pcs/cluster_controller.py (hunk 6 above)
Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py]: Done

File: ha/core/controllers/pcs/cluster_controller.py (hunk 7 above)
Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py]: As discussed over code, I did the code changes

**Review by indrajitzagade — APPROVED** [review]

File: ha/core/controllers/pcs/cluster_controller.py [hunk 9]
```diff
@@ -38,7 +72,37 @@ def start(self) -> dict:
...
+
+                for _node_subgroup in _node_group:
+                    for _node_id in _node_subgroup:
+                        _res = self._controllers[const.NODE_CONTROLLER].start(_node_id)
+                        _res = json.loads(_res)
+                        if _res.get("status") == const.STATUSES.FAILED.value:
+                            msg = _res.get("msg")
+                            Log.error(f"Node {_node_id} : {msg}")
+                    # Wait till all the resources get started in the sub group
+                    time.sleep(const.BASE_WAIT_TIME * const.PCS_NODE_START_GROUP_SIZE)
```
Comment (indrajitzagade) [inline ha/core/controllers/pcs/cluster_controller.py:97]: So we will wait for base time * no. of nodes. But nodes in sub-group will be started parallel, right?

Comment (VijayThakkar132) [reply ha/core/controllers/pcs/cluster_controller.py:97]: Yes

**Review by ajaysrivas — APPROVED** [review]

File: ha/core/controllers/pcs/cluster_controller.py (hunk 9 above)
Comment (ajaysrivas) [reply ha/core/controllers/pcs/cluster_controller.py:97]: These nodes are part of the cluster. While starting the services, they are exchanging messages and this is the reason starting multiple nodes in parallel takes time. This needs to be tuned. Currently, we will use this formula and later we will see if refinement is required.

Comment (indrajitzagade) [reply ha/core/controllers/pcs/cluster_controller.py:97]: got it!

**Pull request merged**